### PR TITLE
Dev/loadmods

### PR DESCRIPTION
**Pull request by RedBrumbler**
- Rename `load_mods` to `load_early_mods` to better reflect what it does
 - Add opening of mods (aka late mods) after the unity method call of `UnityPostLoadApplication`
 - Add load call of those mods right after dlopening them
 - Add `late_load` entry point for early mods and mods, called after mods are dlopened. For mods there is basically no difference between `load` and `late_load`, since they are executed directly after one another. mods just get `late_load` called in order to reduce friction for mod developers

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -179,10 +220,17 @@ MODLOADER_FUNC void modloader_accept_unity_handle([[maybe_unused]] JNIEnv* env,
...
   LOG_DEBUG("Found il2cpp_init at: {}", fmt::ptr(il2cpp_init));
-  modloader_unity_handle = unityHandle;
   // Hook il2cpp_init to redirect to us after loading, and load normal mods
   // TODO: We KNOW it has enough space for us, but we could technically double check this too.
   install_load_hook(reinterpret_cast<uint32_t*>(il2cpp_init));
+
+  void* unity_playerstartfirstscene = dlsym(modloader_unity_handle, "PlayerStartFirstScene");
+  // Hook PlayerStartFirstScene so we can load in mods (aka late mods)
+  // this method is ran to load the first scene of the game (null -> QuestInit for beat saber), and thus after this we are sure unity is initialized
+  // TODO: should we hook UnityInitApplication instead? there is no guarantee that things would work correctly that early, since no scene would be loaded yet
```
Comment (RedBrumbler) [inline src/main.cpp]: Feedback request: we could hook UnityInitApplication instead, but at that point no scene exists yet so GameObjects created at that point couldn't possibly belong to *any* scene.

File: src/main.cpp [hunk 2]
```diff
@@ -59,14 +59,54 @@ void install_load_hook(uint32_t* target) {
...
+  auto* page_aligned_target = reinterpret_cast<uint32_t*>(reinterpret_cast<uint64_t>(target) & ~(kPageSize - 1));
+  FLAMINGO_DEBUG("Marking target: {} as writable, page aligned: {}", fmt::ptr(target), fmt::ptr(page_aligned_target));
+  if (::mprotect(page_aligned_target, kPageSize, PROT_READ | PROT_WRITE | PROT_EXEC) != 0) {
+    // Log error on mprotect!
+    FLAMINGO_ABORT("Failed to mark: {} (page aligned: {}) as +rwx. err: {}", fmt::ptr(target),
+                   fmt::ptr(page_aligned_target), std::strerror(errno));
+  }
+  static flamingo::Trampoline target_hook(target, hookSize, trampoline_size);
+  // PlayerStartFirstScene has a bool parameter, my guess is that it is like the SceneManager::LoadFirstScene, where the bool param determines whether the load occurs as async
+  auto unity_hook = [](bool param) noexcept{
```
Comment (RedBrumbler) [inline src/main.cpp]: Help request: At the moment this does not unhook.
 - Should it?
 - Does it matter? in theory PlayerStartFirstScene would only ever be called once is the assumption

**Comment by hardcpp** [comment]
We might need to look into protecting back memory in it's original state, and also maybe flushing cpu cache

File: shared/loader.hpp [hunk 3]
```diff
@@ -216,6 +220,17 @@ struct LoadedMod {
     }
     return false;
   }
+
+  /// @brief Calls the late_load function on the mod
+  /// @return true if thje call exists and was called, false otherwise
```
Comment (sc2ad) [inline shared/loader.hpp]: nit: typo

File: src/main.cpp [hunk 4]
```diff
@@ -56,17 +57,197 @@ void install_load_hook(uint32_t* target) {
...
+  auto movSz = cs::findNth<1, &movzFind, &cs::insnMatch<>, 1>(*registerNatives);
+  RET_NULL_LOG_UNLESS(movSz);
+  auto ins = **movSz;
+  auto methsz = (ins >> 5) & 0xffff;
+
+  // 0x19 was the size of this array in the unstripped libunity we inject into beat saber.
+  // this is not guaranteed to be the size across every libunty ever,
+  // therefore it not matching is not cause for erroring out of the xref trace
+  if (methsz != 0x19) {
+    LOG_ERROR("Method size found was {} (0x{:x}), while 25 (0x19) was expected", methsz, methsz);
```
Comment (sc2ad) [inline src/main.cpp]: This should warn instwad

**Review by sc2ad — APPROVED** [review]

**Comment by sc2ad** [comment]
I'd like the prune the capstone utils a bit, as we don't need all of that and to remove the cs handle creation there and use flamingo's instead (as we already have an allocated handle). Ideally we also reduce code duplication across usage of flamingo

File: shared/loader.hpp (hunk 3 above)
Comment (RedBrumbler) [reply shared/loader.hpp]: fixed

File: src/main.cpp (hunk 4 above)
Comment (RedBrumbler) [reply src/main.cpp]: Fixed

**Comment by RedBrumbler** [comment]
yeah I probably didn't do a very good job of reducing the utils, mostly just wanted to get the changes pushed so it could be looked at

**Comment by RedBrumbler** [comment]
So after some testing we found that the current code would only work for unity 2021, since the target of the unity hook changes depending on if you're on the 2019 or 2021 unity. (the bl lookup of 20 would be 16 on 2019) this should be solved before this is merged

**Pull request merged**